src/mlpack/methods/lmnn/lmnn_function_impl.hpp

```diff
@@ -0,0 +1,343 @@
+/**
+ * @file lmnn_function_impl.cpp
+ * @author Manish Kumar
+ *
+ * An implementation of the LMNNFunction class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_LMNN_FUNCTION_IMPL_HPP
+#define MLPACK_METHODS_LMNN_FUNCTION_IMPL_HPP
+
+#include "lmnn_function.hpp"
+#include "constraints.hpp"
+
+#include <mlpack/core/math/make_alias.hpp>
+#include <mlpack/core/optimizers/function.hpp>
+
+namespace mlpack {
+namespace lmnn {
+
+template<typename MetricType>
+LMNNFunction<MetricType>::LMNNFunction(const arma::mat& dataset,
+                           const arma::Row<size_t>& labels,
+                           size_t k,
+                           double regularization,
+                           MetricType metric) :
+    dataset(math::MakeAlias(const_cast<arma::mat&>(dataset), false)),
+    labels(math::MakeAlias(const_cast<arma::Row<size_t>&>(labels), false)),
+    k(k),
+    metric(metric),
+    regularization(regularization)
+{
+  // Initialize the initial learning point.
+  initialPoint.eye(dataset.n_rows, dataset.n_rows);
+
+  // Initialize target neighbors.
+  Constraints constraint(dataset, labels, k);
+  constraint.TargetNeighbors(targetNeighbors);
+}
+
+//! Shuffle the dataset.
+template<typename MetricType>
+void LMNNFunction<MetricType>::Shuffle()
+{
+  arma::mat newDataset;
+  arma::Row<size_t> newLabels;
+
+  math::ShuffleData(dataset, labels, newDataset, newLabels);
+
+  math::ClearAlias(dataset);
+  math::ClearAlias(labels);
+
+  dataset = std::move(newDataset);
+  labels = std::move(newLabels);
```

Comment:
Each time the dataset is shuffled, I believe we have to recalculate `TargetNeighbors()`.

Reply:
Or, on second thought, it's not necessary to recalculate the target neighbors, but we do need to apply the same shuffling to the indices.  I guess an easy way to do it for testing is just to call `TargetNeighbors()` again with the shuffled data (but that is very slow and we shouldn't use it as a long-term solution).

Reply:
I read shuffleData() code and I think instead of directly calling shuffleData() we can use its internal code ( as it will give us the shuffle ordering ) and just re-map target neighbors to ordering indices. What do you think?

Reply:
I think using variadic won't work ---- here in case of target neighbors, all the target neighbors will have different indexes (resulting in a change of all row elements too). So, just changing order of target neighbors matrix cols according to shuffling order won't result in correct target neighbors matrix.